common/lib/xmodule/xmodule/js/src/video/09_completion.js

```diff
@@ -0,0 +1,173 @@
+(function(define) {
+    'use strict';
+    /**
+     * Completion handler
+     * @exports video/09_completion.js
+     * @constructor
+     * @param {Object} state The object containing the state of the video
+     * @return {jquery Promise}
+     */
+    define('video/09_completion.js', [], function() {
+        var VideoCompletionHandler = function(state) {
+            if (!(this instanceof VideoCompletionHandler)) {
+                return new VideoCompletionHandler(state);
+            }
+            this.state = state;
+            this.state.completionHandler = this;
+            this.initialize();
+            return $.Deferred().resolve().promise();
+        };
+
+        VideoCompletionHandler.prototype = {
+
+            /** Tears down the VideoCompletionHandler.
+             *
+             *  * Removes backreferences from this.state to this.
+             *  * Turns off signal handlers.
+             */
+            destroy: function() {
+                this.el.remove();
+                this.el.off('timeupdate.completion');
+                this.el.off('ended.completion');
+                delete this.state.completionHandler;
+            },
+
+            /** Initializes the VideoCompletionHandler.
+             *
+             *  This sets all the instance variables needed to perform
+             *  completion calculations.
+             */
+            initialize: function() {
+                this.lastSentTime = undefined;
```

Comment:
This is where I had expected a comment.  Maybe you added them elsewhere?

I don't even know if my comment is accurate, but it would be nice to see something like:
```
// The variables that end with 'Time' are actually in seconds from 0, representing the actual start of the video.
```

Reply:
Done.

Reply:
Thanks.  The one that is confusing to me is what startTime represents.  Is that the startTime of real content, before which if you don't get there it won't be marked complete?  It's up to you whether not you enhance comments.  I'm just interested.

Reply:
You can actually define an edx video that only includes a subset of the source video, so if you have a 10 minute youtube video, you can include only the slice from 8m0s to 9m0s minute of the video by assigning your block a startTime of 480 (`8 * 60`), and an endTime of 540 (`9 * 60`).  So startTime is the time from the start of the source video that the video block will use as its start time.  There's a video in the demo course of a young man reading an excerpt from the Odyssey (or was it the Iliad?) that uses this feature, if you want to see it in action.

Reply:
Thanks for explaining that for me.  That makes more sense than what I was imagining.